File: clients/swift/CryptaliasClient.swift
import Foundation

public enum CryptaliasError: Error {
  case invalidAlias
  case missingResolver
  case missingKey
  case invalidJws
  case missingAddress
  case requestFailed(Int, String)
}

public func resolveAddress(ticker: String, alias: String) async throws -> String {
  if ticker.isEmpty || alias.isEmpty {
    throw CryptaliasError.invalidAlias
  }
  let domain = try parseDomain(alias)
  let configUrl = URL(string: "https://\(domain)/.well-known/cryptalias/configuration")!

  let configData = try await httpGet(url: configUrl, accept: "application/json")
  let configJson = try JSONSerialization.jsonObject(with: configData) as? [String: Any]
  let resolver = ((configJson?["resolver"] as? [String: Any])?["resolver_endpoint"] as? String ?? "").trimmingCharacters(in: CharacterSet(charactersIn: "/"))
  guard !resolver.isEmpty else { throw CryptaliasError.missingResolver }
  guard let jwk = configJson?["key"] as? [String: Any] else { throw CryptaliasError.missingKey }

  let resolveUrl = URL(string: "\(resolver)/_cryptalias/resolve/\(urlEncode(ticker))/\(urlEncode(alias))")!
  let jwsData = try await httpGet(url: resolveUrl, accept: "application/jose")
  let jws = String(decoding: jwsData, as: UTF8.self)
  let payload = try verifyJwsPayload(jws: jws, jwk: jwk)

  guard let address = payload["address"] as? String, !address.isEmpty else {
    throw CryptaliasError.missingAddress
  }
  guard let expires = payload["expires"] as? String else {
    throw CryptaliasError.invalidJws
  }
  try enforceExpires(expires)
  return address
}

public func verifyJwsPayload(jws: String, jwk: [String: Any]) throws -> [String: Any] {
  let parts = jws.split(separator: ".")
  guard parts.count == 3 else { throw CryptaliasError.invalidJws }

  guard let x = jwk["x"] as? String else { throw CryptaliasError.invalidJws }
  let publicKeyData = try base64UrlDecode(x)
  let publicKey = try Curve25519.Signing.PublicKey(rawRepresentation: publicKeyData)

  let signingInput = "\(parts[0]).\(parts[1])"
  let signature = try base64UrlDecode(String(parts[2]))
  let ok = publicKey.isValidSignature(signature, for: Data(signingInput.utf8))
  if !ok {
    throw CryptaliasError.invalidJws
  }

  let payloadData = try base64UrlDecode(String(parts[1]))
  let obj = try JSONSerialization.jsonObject(with: payloadData)
  return obj as? [String: Any] ?? [:]
}

private func parseDomain(_ alias: String) throws -> String {
  guard let idx = alias.lastIndex(of: "$"), idx < alias.index(before: alias.endIndex) else {
    throw CryptaliasError.invalidAlias
  }
  return String(alias[alias.index(after: idx)...])
}

private func httpGet(url: URL, accept: String) async throws -> Data {
  var request = URLRequest(url: url)
  request.setValue(accept, forHTTPHeaderField: "Accept")
  let (data, response) = try await URLSession.shared.data(for: request)
  if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
    let body = String(decoding: data, as: UTF8.self)
    throw CryptaliasError.requestFailed(http.statusCode, body)
  }
  return data
}

private func base64UrlDecode(_ input: String) throws -> Data {
  var b64 = input.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
  while b64.count % 4 != 0 { b64.append("=") }
  if let data = Data(base64Encoded: b64) {
    return data
  }
  throw CryptaliasError.invalidJws
}

private func enforceExpires(_ value: String) throws {
  let formatter = ISO8601DateFormatter()
  formatter.formatOptions = [.withInternetDateTime]
  guard let date = formatter.date(from: value) else {
    throw CryptaliasError.invalidJws
  }
  if date <= Date() {
    throw CryptaliasError.invalidJws
  }
}

private func urlEncode(_ value: String) -> String {
  var allowed = CharacterSet.urlPathAllowed
  allowed.remove(charactersIn: "/")
  return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
}

File: clients/cpp/cryptalias_client.cpp
#include "cryptalias_client.hpp"

#include <curl/curl.h>
#include <nlohmann/json.hpp>
#include <openssl/evp.h>

#include <cctype>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {
size_t write_cb(void* contents, size_t size, size_t nmemb, void* userp) {
  size_t total = size * nmemb;
  auto* buffer = static_cast<std::string*>(userp);
  buffer->append(static_cast<char*>(contents), total);
  return total;
}

std::string http_get(const std::string& url, const std::string& accept) {
  CURL* curl = curl_easy_init();
  if (!curl) {
    throw std::runtime_error("failed to init curl");
  }

  std::string response;
  struct curl_slist* headers = nullptr;
  std::string accept_header = "Accept: " + accept;
  headers = curl_slist_append(headers, accept_header.c_str());

  curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

  CURLcode res = curl_easy_perform(curl);
  long code = 0;
  curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &code);

  curl_slist_free_all(headers);
  curl_easy_cleanup(curl);

  if (res != CURLE_OK) {
    throw std::runtime_error(curl_easy_strerror(res));
  }
  if (code < 200 || code >= 300) {
    throw std::runtime_error("request failed " + std::to_string(code) + ": " + response);
  }
  return response;
}

std::string parse_domain(const std::string& alias) {
  auto pos = alias.rfind('$');
  if (pos == std::string::npos || pos == alias.size() - 1) {
    throw std::runtime_error("alias must be in the format [ticker:]alias$domain");
  }
  return alias.substr(pos + 1);
}

std::string normalize_ticker(const std::string& value) {
  std::string out;
  out.reserve(value.size());
  for (unsigned char c : value) {
    if (!std::isspace(c)) {
      out.push_back(static_cast<char>(std::tolower(c)));
    }
  }
  return out;
}

std::string parse_ticker_prefix(const std::string& alias) {
  auto pos = alias.rfind('$');
  if (pos == std::string::npos || pos == alias.size() - 1) {
    throw std::runtime_error("alias must be in the format [ticker:]alias$domain");
  }
  const std::string left = alias.substr(0, pos);
  auto colon = left.find(':');
  if (colon == std::string::npos) {
    return "";
  }
  if (colon == 0 || colon == left.size() - 1 || left.find(':', colon + 1) != std::string::npos) {
    throw std::runtime_error("invalid format (expected [ticker:]alias[+tag]$domain)");
  }
  return normalize_ticker(left.substr(0, colon));
}

std::string url_encode(const std::string& value) {
  static const char hex[] = "0123456789ABCDEF";
  std::string out;
  for (unsigned char c : value) {
    if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
      out.push_back(static_cast<char>(c));
    } else {
      out.push_back('%');
      out.push_back(hex[c >> 4]);
      out.push_back(hex[c & 0x0F]);
    }
  }
  return out;
}

std::vector<unsigned char> base64url_decode(const std::string& input) {
  std::string b64 = input;
  for (char& c : b64) {
    if (c == '-') c = '+';
    else if (c == '_') c = '/';
  }
  while (b64.size() % 4 != 0) {
    b64.push_back('=');
  }

  static const int kDecTable[256] = {
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,62,-1,-1,-1,63,
      52,53,54,55,56,57,58,59,60,61,-1,-1,-1,-2,-1,-1,
      -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,10,11,12,13,14,
      15,16,17,18,19,20,21,22,23,24,25,-1,-1,-1,-1,-1,
      -1,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,
      41,42,43,44,45,46,47,48,49,50,51,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
      -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1
  };

  std::vector<unsigned char> out;
  int val = 0, valb = -8;
  for (unsigned char c : b64) {
    int d = kDecTable[c];
    if (d == -1) continue;
    if (d == -2) break;
    val = (val << 6) + d;
    valb += 6;
    if (valb >= 0) {
      out.push_back(static_cast<unsigned char>((val >> valb) & 0xFF));
      valb -= 8;
    }
  }
  return out;
}

void enforce_expires(const std::string& value) {
  if (value.empty()) {
    throw std::runtime_error("missing expires in JWS payload");
  }
  std::tm tm = {};
  std::istringstream ss(value);
  ss >> std::get_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
  if (ss.fail()) {
    throw std::runtime_error("invalid expires in JWS payload");
  }
  time_t exp = timegm(&tm);
  if (exp <= time(nullptr)) {
    throw std::runtime_error("resolved address has expired");
  }
}

} // namespace

namespace cryptalias {
std::string resolve_address(const std::string& ticker, const std::string& alias) {
  if (ticker.empty() || alias.empty()) {
    throw std::runtime_error("ticker and alias are required");
  }

  const std::string ticker_clean = normalize_ticker(ticker);
  const std::string prefix = parse_ticker_prefix(alias);
  if (!prefix.empty() && prefix != ticker_clean) {
    throw std::runtime_error("ticker prefix \"" + prefix + "\" does not match \"" + ticker_clean + "\"");
  }
  const std::string domain = parse_domain(alias);
  const std::string cfg_url = "https://" + domain + "/.well-known/cryptalias/configuration";

  const std::string cfg_body = http_get(cfg_url, "application/json");
  auto cfg = nlohmann::json::parse(cfg_body);

  std::string resolver = cfg["resolver"]["resolver_endpoint"].get<std::string>();
  while (!resolver.empty() && resolver.back() == '/') {
    resolver.pop_back();
  }
  if (resolver.empty()) {
    throw std::runtime_error("missing resolver_endpoint in configuration");
  }
  if (!cfg.contains("key")) {
    throw std::runtime_error("missing key in configuration");
  }

  const std::string resolve_url = resolver + "/_cryptalias/resolve/" + url_encode(ticker_clean) + "/" + url_encode(alias);
  const std::string jws = http_get(resolve_url, "application/jose");
  const std::string payload_json = verify_jws_payload(jws, cfg["key"].dump());
  auto payload = nlohmann::json::parse(payload_json);

  if (!payload.contains("address") || payload["address"].get<std::string>().empty()) {
    throw std::runtime_error("missing address in JWS payload");
  }
  if (!payload.contains("expires") || payload["expires"].get<std::string>().empty()) {
    throw std::runtime_error("missing expires in JWS payload");
  }
  enforce_expires(payload["expires"].get<std::string>());
  return payload["address"].get<std::string>();
}

std::string verify_jws_payload(const std::string& jws, const std::string& jwk_json) {
  auto jwk = nlohmann::json::parse(jwk_json);
  if (!jwk.contains("x")) {
    throw std::runtime_error("missing jwk x");
  }
  auto parts_pos1 = jws.find('.');
  auto parts_pos2 = jws.find('.', parts_pos1 + 1);
  if (parts_pos1 == std::string::npos || parts_pos2 == std::string::npos) {
    throw std::runtime_error("invalid JWS format");
  }

  std::string header_b64 = jws.substr(0, parts_pos1);
  std::string payload_b64 = jws.substr(parts_pos1 + 1, parts_pos2 - parts_pos1 - 1);
  std::string sig_b64 = jws.substr(parts_pos2 + 1);
  std::string signing_input = header_b64 + "." + payload_b64;

  auto pub_bytes = base64url_decode(jwk["x"].get<std::string>());
  auto sig_bytes = base64url_decode(sig_b64);

  EVP_PKEY* pkey = EVP_PKEY_new_raw_public_key(EVP_PKEY_ED25519, nullptr, pub_bytes.data(), pub_bytes.size());
  if (!pkey) {
    throw std::runtime_error("failed to create public key");
  }

  EVP_MD_CTX* ctx = EVP_MD_CTX_new();
  if (!ctx) {
    EVP_PKEY_free(pkey);
    throw std::runtime_error("failed to create md ctx");
  }

  int ok = EVP_DigestVerifyInit(ctx, nullptr, nullptr, nullptr, pkey);
  if (ok != 1) {
    EVP_MD_CTX_free(ctx);
    EVP_PKEY_free(pkey);
    throw std::runtime_error("verify init failed");
  }
  ok = EVP_DigestVerify(ctx, sig_bytes.data(), sig_bytes.size(),
                        reinterpret_cast<const unsigned char*>(signing_input.data()), signing_input.size());
  EVP_MD_CTX_free(ctx);
  EVP_PKEY_free(pkey);

  if (ok != 1) {
    throw std::runtime_error("signature verification failed");
  }

  auto decoded = base64url_decode(payload_b64);
  return std::string(decoded.begin(), decoded.end());
}

} // namespace cryptalias
